api/envoy/extensions/filters/http/api_key_auth/v3/api_key_auth.proto

```diff
@@ -17,38 +16,84 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 option (xds.annotations.v3.file_status).work_in_progress = true;
 
 // [#protodoc-title: APIKey Auth]
-// [#not-implemented-hide:]
 // APIKey Auth :ref:`configuration overview <config_http_filters_api_key_auth>`.
 // [#extension: envoy.filters.http.api_key_auth]
 
 // API Key HTTP authentication.
 //
-// Example:
+// For example, the following configuration configures the filter to authenticate the clients using
+// the API key from the header ``X-API-KEY``. And only the clients with the key ``real-key`` are
+// considered as authenticated.
 //
 // .. code-block:: yaml
 //
-//    authentication_header: "X-API-KEY"
-//    keys:
-//      inline_string: |-
-//        clientID1:apiKey1
-//        clientID2:apiKey2
+//    credentials:
+//    - key: real-key
+//      client: user
+//    key_sources:
+//     - header: "X-API-KEY"
 //
-message APIKeyAuth {
-  // keys used to authenticate the client.
-  // It should be a map of clientID to apiKey.
-  // The clientID serves solely for identification purposes and isn't used for authentication.
-  config.core.v3.DataSource keys = 1 [(udpa.annotations.sensitive) = true];
+message ApiKeyAuth {
+  // The credentials that are used to authenticate the clients.
+  repeated Credential credentials = 1 [(udpa.annotations.sensitive) = true];
+
+  // The key sources to fetch the key from the coming request.
+  repeated KeySource key_sources = 2;
+}
+
+// API key auth configuration of per route or per virtual host or per route configuration.
+message ApiKeyAuthPerRoute {
+  // Route specific APIKeyAuth configuration. This is optional and could be used to **override** the
+  // filter level ApiKeyAuth configuration **partly**.
+  //
+  // For example, if non-empty ``credentials`` is set in the ``override_config``, then the
+  // ``credentials`` in the filter level configuration will be ignored and the ``credentials`` in
+  // this configuration will be used.
```

Comment:
I don't think partial override is the right way to go. If a message of type `ApiKeyAuth` is provided, then it should replace the one define above entirely. The reason I'm against this is that if a new field is added in the future to `ApiKeyAuth` where it won't make sense to do a partial override, this will complicate the use of this field and the override.

If an override to a specific field is desired, then an additional field can be added to `ApiKeyAuthPerRoute`. For example `repeated Credential credentials` that will override only the credentials defined above.